packages/next/next-server/server/render.tsx

```diff
@@ -507,7 +507,19 @@ export async function renderToHTML(
   await Loadable.preloadAll() // Make sure all dynamic imports are loaded
 
   // url will always be set
-  const asPath: string = renderOpts.resolvedAsPath || (req.url as string)
+  let asPath: string = renderOpts.resolvedAsPath || (req.url as string)
+
+  if (!renderOpts.nextExport && isSSG) {
+    // this does not decode query values only the path
+    // ensuring non-ascii values are the proper encoding
+    // and ensures we match prerender asPath values
+    try {
+      asPath = decodeURI(asPath)
+    } catch (_) {
+      // non-fatal continue
+    }
```

Comment:
Swallowing this concerns me.

Reply:
Was going back and fourth on whether this should be a 400 error if it fails to decode to match the dynamic-route decode error handling, do you think that would be more correct here?

Reply:
Considering that something should always be in an encoded or decoded state already, this should probably 400.

Reply:
This is now removed and when decoding fails while normalizing the `ssgCacheKey` in `next-server` we now throw a `DECODE_ERROR` which triggers a 400 response.